R/src_gen.R

```diff
@@ -0,0 +1,207 @@
+#' @title Generate Synthetic Rating Curve
+#' @param comids A set of COMIDs. See \link{find_comids}.
+#' @param hand A HAND \code{raster}.
+#' @param stage A set of stage values. Defaults to \code{0:20}.
+#' @param progress If \code{TRUE}, output a progress bar.
+#' @return A \code{data.frame} representing the SRC table.
+#' @importFrom nhdplusTools get_vaa
+#' @importFrom data.table :=
+#' @importFrom plyr .
+#' @export
+get_src <- function(comids, hand, stage = 0:20, progress = TRUE) {
+    if (progress) {
+        pb <- progress::progress_bar$new(
+            total = length(stage) + 3,
+            format = " generating SRC [:bar] :percent eta: :eta"
+        )
+
+        pb$tick(0)
+    }
+
+    if (missing(comids) & !missing(hand)) {
+        hand_bb <- sf::st_bbox(hand) %>%
+                   sf::st_as_sfc() %>%
+                   sf::st_as_sf()
+
+        comids <- find_comids(hand_bb)
+    }
+
+    # If the HAND raster is not passed, get from AWS
+    if (missing(hand) & !missing(comids)) {
+        nhd <- nhdplusTools::get_nhdplus(comid = comids)
+        hand <- get_hand_raster(nhd)
+    }
+
+    # Fix NA values
+    hand[is.na(hand[])] <- 0
+    # Generate Slope Raster
+    slope <- raster::terrain(hand)
+    slope[is.na(slope[])] <- 0
+    # Get raster cell resolution
+    res <- compute_cellres(hand)
+    # Create HAND_Slope data frame
+
+    if (progress) pb$tick()
+
+    hand_slope <-
+        data.frame(
+            hand = raster::getValues(hand),
+            slope = raster::getValues(slope)
+        ) %>%
+        na.omit() %>%
+        dplyr::arrange(hand) %>%
+        dplyr::mutate(
+            res = res,
+            BA  = res * res * sqrt(1 + slope)
+        )
+
+    if (progress) pb$tick()
+
+    # Get VAAs via nhdplusTools
+    vaa <-
+        nhdplusTools::get_vaa(
+            atts = c(
+                "pathlength",
+                "arbolatesu",
+                "lengthkm",
+                "areasqkm",
+                "slope",
+                "roughness"
+            )
+        ) %>%
+        dplyr::filter(comid %in% comids)
+
+    if (progress) pb$tick()
+
+    synthetic_rating_curve <-
+        lapply(
+            stage,
+            FUN = function(y) {
+
+                if (progress) pb$tick()
+
+                vaa %>%
+                    na.omit() %>%
```

Comment:
Move to initail creation of `vaa`,  na.omit can be slow in large loop (or apply)

Reply:
Done. Will show on next commit 👍